protocols/autonatv2/src/client/handler/dial_request.rs

```diff
@@ -0,0 +1,396 @@
+use futures::{
+    channel::{mpsc, oneshot},
+    AsyncRead, AsyncWrite, SinkExt, StreamExt,
+};
+use futures_bounded::FuturesSet;
+use libp2p_core::{
+    upgrade::{DeniedUpgrade, ReadyUpgrade},
+    Multiaddr,
+};
+
+use libp2p_identity::PeerId;
+use libp2p_swarm::{
+    handler::{
+        ConnectionEvent, DialUpgradeError, FullyNegotiatedOutbound, OutboundUpgradeSend,
+        ProtocolsChange,
+    },
+    ConnectionHandler, ConnectionHandlerEvent, Stream, StreamProtocol, StreamUpgradeError,
+    SubstreamProtocol,
+};
+use std::{
+    collections::VecDeque,
+    convert::identity,
+    io,
+    iter::{once, repeat},
+    sync::Arc,
+    task::{Context, Poll},
+    time::Duration,
+};
+
+use crate::client::behaviour::Error;
+use crate::{
+    client::behaviour::Event,
+    generated::structs::{mod_DialResponse::ResponseStatus, DialStatus},
+    protocol::{
+        Coder, DialDataRequest, DialDataResponse, DialRequest, DialResponse, Request, Response,
+        DATA_FIELD_LEN_UPPER_BOUND, DATA_LEN_LOWER_BOUND, DATA_LEN_UPPER_BOUND,
+    },
+    REQUEST_PROTOCOL_NAME,
+};
+
+#[derive(Debug, thiserror::Error)]
+pub(crate) enum InternalError {
+    #[error("io error")]
+    Io(#[from] io::Error),
+    #[error("invalid referenced address index: {index} (max number of addr: {max})")]
+    InvalidReferencedAddress { index: usize, max: usize },
+    #[error("data request too large: {len} (max: {max})")]
+    DataRequestTooLarge { len: usize, max: usize },
+    #[error("data request too small: {len} (min: {min})")]
+    DataRequestTooSmall { len: usize, min: usize },
+    #[error("server rejected dial request")]
+    ServerRejectedDialRequest,
+    #[error("server chose not to dial any provided address")]
+    ServerChoseNotToDialAnyAddress,
+    #[error("server ran into an internal error")]
+    InternalServer,
+    #[error("server did not respond correctly to dial request")]
+    InvalidResponse,
+    #[error("server was unable to connect to address: {addr:?}")]
+    UnableToConnectOnSelectedAddress { addr: Option<Multiaddr> },
+    #[error("server experienced failure during dial back on address: {addr:?}")]
+    FailureDuringDialBack { addr: Option<Multiaddr> },
+}
+
+#[derive(Debug)]
+pub struct TestEnd {
+    pub(crate) dial_request: DialRequest,
+    pub(crate) reachable_addr: Multiaddr,
+}
+
+#[derive(Debug)]
+pub enum ToBehaviour {
+    TestCompleted(Result<TestEnd, Error>),
+    StatusUpdate(Event),
+    PeerHasServerSupport,
+}
+
+pub struct Handler {
+    queued_events: VecDeque<
+        ConnectionHandlerEvent<
+            <Self as ConnectionHandler>::OutboundProtocol,
+            <Self as ConnectionHandler>::OutboundOpenInfo,
+            <Self as ConnectionHandler>::ToBehaviour,
+        >,
+    >,
+    outbound: futures_bounded::FuturesSet<Result<TestEnd, crate::client::behaviour::Error>>,
```

Comment:
```suggestion
    outbound: futures_bounded::FuturesSet<Result<TestEnd, Error>>,
```

This is already imported.

Reply:
That changed anyway so it doesn't matter.